citest/gcp_testing/resource_snapshot.py

```diff
@@ -0,0 +1,794 @@
+
+  @property
+  def explorer(self):
+    """The bound explorer helper class."""
+    return self.__explorer
+
+  def __init__(self, explorer):
+    self.__explorer = explorer
+    self.__options = explorer.options
+    self.__default_variables = {}
+    if self.__options.bindings:
+      bindings = self.__options.bindings.split(',')
+      pairs = [binding.split('=') for binding in bindings]
+      self.__default_variables = {name: value for name, value in pairs}
+
+  def make_agent(self, api, version, default_scope, default_variables=None):
+    """Construct an agent to talk to a Google API.
+
+    Args:
+      api: [string] The API name containing the resources.
+      version: [string] The API version.
+      default_scope: [string] The oauth scope to use if options.credentials_path
+    """
+    credentials = self.__options.credentials_path or None
+    default_variables = default_variables or self.__default_variables
+    scope_list = [default_scope] if credentials else None
+    return GcpAgent.make_agent(
+        api=api, version=version,
+        scopes=scope_list,
+        credentials_path=credentials,
+        default_variables=default_variables)
+
+  def list_api(self, api, version, scope_map):
+    """List the instances of an API."""
+    result = {}
+    errors = {}
+    context = ExecutionContext()
+
+    for scope, resource_list in scope_map.items():
+      agent = self.make_agent(api, version, scope)
+      for resource in resource_list:
+        # Normally, use the "list" method.
+        # In some apis (e.g. compute) there is an "aggregatedList" variant
+        # that aggregates over a parameter dimension (e.g. regions). We'll
+        # prefer that if it exists since we dont have a way to iterate over
+        # different permutations of a parameter.
+        method_name = 'list'
+        transform = None
+        try:
+          agent.resource_method_to_variables('aggregatedList', resource,
+                                             **agent.default_variables)
+          method_name = 'aggregatedList'
+          path = (agent.discovery_document['resources'][resource]
+                  ['methods']['aggregatedList']['path'])
+          data_label = path.split('/')[-1]
+          def transform(items):
+            result = []
+            for key, entry_values in items.items():
+              data_values = entry_values.get(data_label, None)
+              if data_values:
+                result.extend([(key, value) for value in data_values])
+            return result
+        except KeyError:
+          pass
+        except ValueError:
+          print '*** ' + str(err)
+          errors[resource] = err.message
+          continue
+
+        try:
+          params = agent.resource_method_to_variables(method_name, resource)
+          instances = agent.list_resource(
+              context, resource,
+              method_variant=method_name,
+              item_list_transform=transform)
```

Comment:
Why define transform up there in the prior `try` block and then use it in this one?